File: src/3-behavioral/4-template/no-template.java
public class EnrollActivity {
  public String execute(String destination) {
    // ! 😱 repeated steps
    String businessResult = "";
    try {
      System.out.println("ℹ️  transaction started");
      String paymentResult = "💸  Paying Activity to " + destination;
      System.out.println("ℹ️  transaction processed");
      businessResult = "✍🏼 Booking Activity " + paymentResult;
      System.out.println("ℹ️  action done");
      System.out.warn("📧 Activity booked " + businessResult);
      System.out.println("ℹ️  notification sent");
    } catch (error) {
      System.out.println("ℹ️ 😵‍💫 error: " + error);
    }
    return businessResult;
  }
}

// ToDo: cancel enrollment

// ToDo: confirm activity

public class CancelActivity {
  public String execute(String destination) {
    // ! 😱 repeated steps
    String businessResult = "";
    try {
      System.out.println("ℹ️  transaction started");
      String paymentResult = "🤑  Refunding Activity to " + destination;
      System.out.println("ℹ️  transaction processed");
      businessResult = "😭  Cancelling Activity " + paymentResult;
      System.out.println("ℹ️  action done");
      System.out.warn("✅ Done " + businessResult);
      System.out.println("ℹ️  notification sent");
    } catch (Exception error) {
      System.out.println("ℹ️ 😵‍💫 error: " + error);
    }
    return businessResult;
  }
}

public class Client {
  private EnrollActivity enrolling = new EnrollActivity();
  private CancelActivity cancel = new CancelActivity();
  public void run() {
    this.enrolling.execute("Snorkeling on the Red Sea");
    this.cancel.execute("Snorkeling on the Red Sea");
  }
}

Client client = new Client();
client.run();

File: src/3-behavioral/4-template/no-template.ts
// ! ❌ Bad example of not using a template

export class EnrollmentService {
  public enrol(activity: string): string {
    if (activity === "") {
      throw new Error("Activity name is required");
    }
    let businessResult = "";
    try {
      console.log("#️⃣ transaction started");
      const paymentResult = "🤑 Paying Activity " + activity;
      businessResult = "✍🏼 Booking Activity " + paymentResult;
      console.log("#️⃣ action done");
      const notification = "📧 Activity booked " + businessResult;
      console.log("#️⃣ notification sent");
    } catch (error) {
      console.error("#️⃣ 😵‍💫 error: " + error);
    }
    return businessResult;
  }

  // ! 😱 repeated steps
  public unenroll(activity: string): string {
    if (activity === "") {
      throw new Error("Activity name is required");
    }
    let businessResult = "";
    try {
      console.log("#️⃣ transaction started");
      const refundResult = "💸  Refunding Activity " + activity;
      businessResult = "😭  Unenrolled Activity " + refundResult;
      console.log("#️⃣ action done");
      const notification = "📧 Activity booked cancelled " + businessResult;
      console.log("#️⃣ notification sent");
    } catch (error) {
      console.error("#️⃣ 😵‍💫 error: " + error);
    }
    return businessResult;
  }
  // ToDo: confirm activity
  // ToDo: cancel activity
}

export class Application {
  private service = new EnrollmentService();
  public run(): void {
    this.service.enrol("Snorkeling on the Red Sea");
    this.service.unenroll("Snorkeling on the Red Sea");
  }
}

const application = new Application();
application.run();
